packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt

```diff
@@ -117,7 +118,35 @@ actual object RealmInterop {
             realmc.classArray_setitem(cclasses, i, cclass)
             realmc.propertyArrayArray_setitem(cproperties, i, classProperties)
         }
-        return LongPointerWrapper(realmc.realm_schema_new(cclasses, count.toLong(), cproperties))
+        try {
+            return LongPointerWrapper(realmc.realm_schema_new(cclasses, count.toLong(), cproperties))
+        } finally {
+            // Clean up classes
+            for (classIndex in 0 until count) {
+                val classInfo = realmc.classArray_getitem(cclasses, classIndex)
+
+                // Clean properties
+                val propertyArray = realmc.propertyArrayArray_getitem(cproperties, classIndex)
+
+                val propertyCount = classInfo.getNum_properties() + classInfo.getNum_computed_properties()
+                for (propertyIndex in 0 until propertyCount) {
+                    val property = realmc.propertyArray_getitem(propertyArray, propertyIndex.toInt())
+
+                    realmc.realm_property_info_t_cleanup(property)
+                    property.delete()
+                }
+
+                realmc.delete_propertyArray(propertyArray)
+
+                // end clean properties
+
+                realmc.realm_class_info_t_cleanup(classInfo)
+                classInfo.delete()
+            }
+
+            realmc.delete_propertyArrayArray(cproperties)
+            realmc.delete_classArray(cclasses)
```

Comment:
Not sure I get why we need to do cleanup this way? Also, this is quite a lot of JNI traversals to do cleanup. Couldn't we move this entire block into a helper function in `realm_api_helpers`?

Reply:
Also, doesn't `realmc.classArray_getitem` cause a new allocation rather than returning the old one?

Reply:
But now that you mention, in `realm_get_class` we could free the allocated struct afterwards.

Reply:
This is still doing a ton of JNI traversals. My comment was about us moving the entire for-loop into C++. I assume that is possible?

Reply:
Yes, it is possible. I wanted to use, if possible, swig APIs, and then in a separate PR move this logic to c++ to measure the actual performance gain from doing it.

Reply:
Ah, fair enough 👍